sql/server_component/mysql_command_backend.cc

```diff
@@ -75,6 +75,21 @@ mysql_state_machine_status cssm_begin_connect(mysql_async_connect *ctx) {
   my_h_service h_command_consumer_srv = nullptr;
 
   MYSQL_SESSION mysql_session = nullptr;
+
+  /* We need to handle the failure in this function.
+     Setting mcs_extn->session_svc right after session open is not enough
+     to handle user lookup errors (and following errors as well)
+     because in case of this function returns error, mysql->extension is
+     cleaned up immediately by the caller. The caller does not take care of
+     session_svc, because it is not aware of this structure.
+  */
+  std::shared_ptr<MYSQL_SESSION> mysql_session_close_guard(
```

Comment:
I was struggling with the same double indirection problem some time ago.
`MYSQL_SESSION` is by itself a pointer to an opaque structure. So, I did this trick with
`std::remove_pointer_t<>`;
https://github.com/percona/percona-server/blob/8.0/components/masking_functions/src/masking_functions/sql_context.cpp#L105
In my opinion this make the code more readable.
The only problem I see if you go my way, you won't be able to do `mysql_session = nullptr`. You will need
to call something like `mysql_session_close_guard.release()` but for this, in turns you will need `std::unique_ptr<>` with more boilerplate instead of `std::shared_ptr<>`.
In any case this is just a butification comment.